MMU3 Feature

eoyilmaz opened the pull request:
### Description

This adds the MMU3 feature from Prusa Firmware.

### Requirements

No special requirements

### Benefits

It is now possible to use MMU3 hardware along with the features like SpoolJoin etc.

### Related Issues

- #26523
- #17735

eoyilmaz committed:
MMU3 Feature

The MMU3 feature from Prusa firmware has been migrated and the changes required for the integration has been introduced.

eoyilmaz committed:
MMU3 Feature

Updated `features.ini` to be able to build with MMU3 feature.

eoyilmaz committed:
Fix for the test build issue.

eoyilmaz committed:
Updated `SpoolJoin::SpoolJoin::toggleSpoolJoin()` to call `settings.save()`

eoyilmaz commented:
Hi,

Sorry for not dropping a detailed PR message earlier today, I didn't have time for that.

So this is the MMU3 feature from PrusaFirmware. I tested the code heavily (albeit with only one printer) and it is working fine (but, I need to do more tests after I rebased the code to Marlin's HEAD to create this PR, I see lot's of things have changed).

There are lots of things that I need to drop some note to. But if you were following the #26523 I already reported mostly about my progress there, and the code explains itself.

For a short explanation, 95% of the code is migrated from PrusaFirmware and because it is mostly based on Marlin1 (if I'm not misttaken) I had to replace most of the code that are interacting with the rest of the Marlin, i.e. creating menus, querying something from the planner, saving something to the EEPROM etc. othwerwise the code is directly transferred from Prusa firmware.

And becasue the code is somewhere in between PrusaFirmware and Marlin I think there are places that we have room for optimization. (just remembered that I was going to update the format to use 2 spaces, will do it in a minute).

Anyways, I'm re-reading the code in this PR and I'll drop notes to places where I think I need supervision.

eoyilmaz committed:
Fix for the test build issues.

eoyilmaz commented:
I working on the build test errors 👍

eoyilmaz committed:
Fix for the test build issues.

eoyilmaz committed:
Fix for the test build issues.

eoyilmaz committed:
Fix for the test build issues for `mega2560` and `rambo`.

eoyilmaz committed:
Added test for rambo board with MMU3 feature.

eoyilmaz committed:
- Fixed build errors for `rambo` board.
- Made `MMU_MENUS` a requirement for `MMU3`.

eoyilmaz committed:
- Fixed build errors for `rambo` board.
- Made `EEPROM_SETTINGS` a requirement for `MMU3`.

eoyilmaz committed:
- Added SKR 3 EZ board tests.

eoyilmaz committed:
- Moved SKR 3 EZ board tests to STM32H743VI_btt board tests.

eoyilmaz committed:
- Fix `test-builds.yml`.

eoyilmaz committed:
- Fix STM32H743VI_btt tests.

eoyilmaz committed:
MMU3:

Updated white space formatting to 2 spaces.

eoyilmaz commented:
Hi there,

I think the code is now to be reviewed...

First, let me say that I'm not a C++ developer, or at least not for the last 10 years or so... So there might be (or definitelly will be) some suboptimal things that I added on top of Prusa's code, especially on the `mmu2_reporting.cpp` I'm very confused with how I need to use the `F()` or `PSTR()` and I was not able to modify the code so that it passes the tests both in `STM32` and `AVR` and I've choosen `STM32` compatible code over the `AVR`.

**Known Problems & Bugs and more things TODO**
- Menus can be finiky. If I don't start the MMU3 before or at the same time the printer starts, I can't select anythign other than the first 2 items in the menus. I was having menu issues with MMU2 too and I'll definitely look in to this.
- I didn't throughly check the M707/M708 commands (I'll look in to them as a follow up).
- Enable/Disable MMU menu item is missing, this is fairly easy, the API is there only the menu items are missing, and I might even do it while you are reviewing the code or I can leave it to another follow up PR.
- `feature/mmu3/mmu_hw/errors_list.h` can definitelly be merged to `language_en.h` but I'm not experienced enough to do it as the Prusa developers are getting the error code from MMU3 as an `uint16_t` and converting it to a `static const char[] PROGMEM` by using a dictionary, there are helper functions to do that... You guys would probably do it in 5 minutes or so...

and I think that's all... I really enjoyed working on this, and I wanted to migrate this code to Marlin at the moment Prusa announced it... I was not sure if I was ever be able to do it, and I'm really proud of myself that I was able to do it and I'm really greateful both to the Prusa and Marlin community and developers that allowed me to run an MMU3 on my Ender 3 Pro... You guys are great 👍

eoyilmaz committed:
- Fixed the default value for `show_lcd` to `true` in `pause::resume_print()` to match the previous implementation.
- Fixed `M600` to show the LCD options, which was broken after MMU3 implementation.

eoyilmaz committed:
Added MMU On/Off menu item to the MMU menus.

eoyilmaz commented:
Hi @thisiskeithb, do you need anything for this to be reviewed/merged?

eoyilmaz committed:
- [MMU3] Added "Reset Fail Stats" menu item.
- [MMU3] Updated Load Filament/Load All Filament/Unload Filament menu actions to fix #24813.
- [MMU3] Added `MMU2::OperationStatistics::reset_fail_stats()` that resets only the fail stats and keeps the material change counter.

eoyilmaz committed:
- [MMU3] Updated `MMU2::OnMMUProgressMsgSame()` so that when loading the filament to the gears the extrusion length is determined through the `logic.PulleySlowFeedRate()` and limited by the `EXTRUDE_MAXLENGTH` setting.

eoyilmaz committed:
[MMU3]

- Updated statistics menu to use longer menu names.
- Updated statistics menu to show the label "Current" instead of "Last" if a print job is currently ongoing.

thisiskeithb commented:
> Hi @thisiskeithb, do you need anything for this to be reviewed/merged?

Thinkyhead will do a review once you're done making changes & he has time. There are quite a few other open PRs and there's still prep going on for the next release.

eoyilmaz committed:
[MMU3] Fixed M707/M708 G-Code commands to properly parse 16bit hex values.

eoyilmaz committed:
- Revert accidental changes to `pins/pins_BTT_SKR_V3_0_common.h`.

eoyilmaz committed:
[MMU3] Updated MMU menu item order to better match Prusa documentation.

rerickson1 commented:
Just some friendly feedback. You shouldn't merge the bugfix-2.1.x branch into this PR. You should be doing a git rebase to rebase your changes onto that branch when you want to "pull in" updates. This PR should only contain your commits for the new feature.

eoyilmaz commented:
> Just some friendly feedback. You shouldn't merge the bugfix-2.1.x branch into this PR. You should be doing a git rebase to rebase your changes onto that branch when you want to "pull in" updates. This PR should only contain your commits for the new feature.

Actually I was trying to do a rebase to update some of my commit messages, and I think I messed up there... I believe I can update the PR to clean it, if desired.

eoyilmaz commented:
> Just some friendly feedback. You shouldn't merge the bugfix-2.1.x branch into this PR. You should be doing a git rebase to rebase your changes onto that branch when you want to "pull in" updates. This PR should only contain your commits for the new feature.

Is there any easy way to fix this at this stage? I tried doing a rebase and force pushed to my branch, but I think it didn't work as we want it to be.

thisiskeithb commented:
> Is there any easy way to fix this at this stage?

I dropped all the extra/duplicate commits & did another force push.

eoyilmaz commented:
> > Is there any easy way to fix this at this stage?
> 
> I dropped all the extra/duplicate commits & did another force push.
> duplicates:

aah thank you @thisiskeithb 👍

thinkyhead commented:
> Is there any easy way to fix this at this stage?

```
git fetch upstream
git merge upstream/bugfix-2.1.x
git reset upstream/bugfix-2.1.x
git commit -m "MMU3 Feature"
```

thinkyhead commented on ini/features.ini:
```diff
@@ -258,6 +258,7 @@ HAS_MEATPACK                           = build_src_filter=+<src/feature/meatpack
 HAS_PRUSA_MMU1                         = build_src_filter=+<src/feature/mmu/mmu.cpp>
 HAS_PRUSA_MMU2                         = build_src_filter=+<src/feature/mmu/mmu2.cpp> +<src/gcode/feature/prusa_MMU2>
+HAS_PRUSA_MMU3                         = build_src_filter=+<src/feature/mmu3/mmu2.cpp> +<src/feature/mmu3/messages.cpp> +<src/feature/mmu3/mmu2_crc.cpp> +<src/feature/mmu3/mmu2_error_converter.cpp> +<src/feature/mmu3/mmu2_fsensor.cpp> +<src/feature/mmu3/mmu2_log.cpp> +<src/feature/mmu3/mmu2_marlin1.cpp> +<src/feature/mmu3/mmu2_power.cpp> +<src/feature/mmu3/mmu2_progress_converter.cpp> +<src/featu ...
 PASSWORD_FEATURE                       = build_src_filter=+<src/feature/password> +<src/gcode/feature/password>
```
If you only want to enable all the `.cpp` files in the `src/feature/mmu3` folder then you only need `+<src/feature/mmu3>` with our latest build scripts.

thinkyhead commented:
The first hour or two of the review and prep time for this PR will be spent adding the standard Marlin header (with © 2024) to all the new source files and reformatting the code according to our standards (i.e., by using the included "uncrust" script and then following up with some adjustments). It would be a great help if you would go ahead and do this yourself so that we don't have to. At that point it will be "ready for review."

rerickson1 commented:
@eoyilmaz FYI, I am in the process of trying out this PR on my printer. Don't let me hold anything up, just wanted to let everyone know.

eoyilmaz commented:
> The first hour or two of the review and prep time for this PR will be spent adding the standard Marlin header (with © 2024) to all the new source files and reformatting the code according to our standards (i.e., by using the included "uncrust" script and then following up with some adjustments). It would be a great help if you would go ahead and do this yourself so that we don't have to. At that point it will be "ready for review."

sure sure, this is my first PR to Marlin in this scale... I'll check the contribution guides again, and please do not spend extra time on aything that I can do 👍

rerickson1 commented on Marlin/Configuration_adv.h:
```diff
@@ -4378,44 +4378,90 @@
+  // ----
+  // These values are compatible with MMU3 as they are defined in mm/s
+
+  #define MMU2_EXTRUDER_PTFE_LENGTH 42.3f // mm
+  #define MMU2_EXTRUDER_HEATBREAK_LENGTH  17.7f // mm
```
Should these values added together represent the distance from the extruder gears to the nozzle tip? Maybe we can clarify this with comments?

eoyilmaz replied:
I think I moved them up from around line 4540 where a comment exists already. But yeah a comment would be more descriptive.

rerickson1 commented on Marlin/Configuration_adv.h:
```diff
@@ -4440,6 +4501,66 @@
+    // Beware - this value is used to initialize the MMU logic layer - it will be sent to the MMU upon line up (written into its 8bit register 0x0b)
+    // However - in the G-code we can get a request to set the extra load distance at runtime to something else (M708 A0xb Xsomething).
+    // The printer intercepts such a call and sets its extra load distance to match the new value as well.
+    #define MMU2_FILAMENT_SENSOR_POSITION 0 // mm
```
Is this the distance from the filament sensor to the gears?

eoyilmaz replied:
Exactly, and it accepts negative values too (considering your setup it will be beneficial for you).

eoyilmaz commented:
> adding the standard Marlin header (with © 2024) to all the new source files and reformatting the code according to our standard

@thinkyhead as I see, the "uncrust" script adds the Marlin header to all the `*.h` files (I'll update to 2024), and it contains similar text to the one that exists in: https://github.com/prusa3d/Prusa-Firmware/blob/MK3/Firmware/Firmware.ino

Do we need to add anything to that as these files are coming from PrusaFirmware? Or is this enough?

eoyilmaz committed:
[MMU3] Addressed review comments.

eoyilmaz commented:
> The first hour or two of the review and prep time for this PR will be spent adding the standard Marlin header (with © 2024) to all the new source files and reformatting the code according to our standards (i.e., by using the included "uncrust" script and then following up with some adjustments). It would be a great help if you would go ahead and do this yourself so that we don't have to. At that point it will be "ready for review."

@thinkyhead all are done 👍

rerickson1 commented on Marlin/src/feature/mmu3/mmu2_reporting.cpp:
```diff
@@ -0,0 +1,794 @@
+ * mmu2_reporting.cpp
+ */
+
+#include <avr/pgmspace.h>
```
I am having issues with this entire PR. It will not compile for non-AVR platforms.
I am trying to compile for board `BOARD_MKS_SGEN_L_V2` (`LPC1769`)

thinkyhead replied:
Hopefully this is resolved by now through our helpful support network (?) If not, we'll have to take a closer look.

thisiskeithb replied:
Lots of stuff got buried:

<img width="227" alt="image" src="https://github.com/MarlinFirmware/Marlin/assets/13375512/73cd365e-ffac-4cbd-a0cf-955c6f08a128">

rerickson1 replied:
This is resolved.

eoyilmaz committed:
[MMU3] Fix "finicky menus" caused by encoder position reset.

eoyilmaz committed:
[MMU3] Fixed `mmu2_protocol_logic` for non-AVR.

rerickson1 commented:
```
In file included from Marlin/src/feature/mmu3/mmu2.h:38,
                 from Marlin/src/MarlinCore.cpp:235:
Marlin/src/feature/mmu3/mmu2_protocol_logic.h:58:14: fatal error: pgmspace.h: No such file or directory

******************************************************************
* Looking for pgmspace.h dependency? Check our library registry!
*
* CLI  > platformio lib search "header:pgmspace.h"
* Web  > https://registry.platformio.org/search?q=header:pgmspace.h
*
******************************************************************

   58 |     #include <pgmspace.h>
      |              ^~~~~~~~~~~~
compilation terminated.
*** [.pio/build/LPC1769/src/src/MarlinCore.cpp.o] Error 1
```

rerickson1 commented:
In case it helps, pushed up my branch: https://github.com/rerickson1/Marlin/commits/rde/2.1.x_mmu3/

eoyilmaz commented:
thank you @rerickson1 I'll look into that later on, currently I'm working on the "full screen error reporting" stuff...

eoyilmaz committed:
[MMU3] Full screen error messages:

Enabled the full screen error messages which allows the user to interact with the MMU directly from the LCD menus and get more info about the error if needed.

eoyilmaz committed:
[MMU3] Fix for failing build tests.

eoyilmaz commented:
Hey @thinkyhead I'm done implementing & fixing the things I planned to implement & fix, and this is now ready to be reviewed.

eoyilmaz commented:
> In case it helps, pushed up my branch: https://github.com/rerickson1/Marlin/commits/rde/2.1.x_mmu3/

@rerickson1 I fetched your branch and created a test case out of it but I have a tone of problems unrelated to MMU3 when building it, were you able to build it at all? Do I need to change anything other than anything in `Configuration.h` and `Configuration_adv.h`.

eoyilmaz commented:
@thinkyhead do you think that you'll be able to review this now?

eoyilmaz committed:
[MMU3] always wait for a request's expected response.

eoyilmaz committed:
[MMU3] Fix nozzle double parking.

In very rare conditions, i.e. a second error arise while trying to fix the first one, even though thee nozzle is already parked, it lifts in the Z axis a second time, causing the nozzle not to be able to return to its original Z height when the errors are solved (there will be an offset of +NOZZLE_PARK_POINT.z). This commit fixes that.

eoyilmaz committed:
[MMU3] Fix nozzle double parking (second attempt).

- Apparently there were more places that the nozzle is lifted in Z unnecessarily.

eoyilmaz committed:
[MMU3] Fix for cutter initiation on `MAX_RETRIES - 1`.

Fixed the `MMU2::ToolChangeCommonOnce()` to properly initiate cutter at the last attempt instead of the hard coded 2nd attempt.

eoyilmaz commented:
@thinkyhead do you think that we can merge this now?

rerickson1 commented:
There are still build issues:
```
In file included from Marlin/src/feature/mmu3/mmu2.h:38,
Compiling .pio/build/LPC1769/src/src/feature/bedlevel/ubl/ubl.cpp.o
                 from Marlin/src/MarlinCore.cpp:235:
Marlin/src/feature/mmu3/mmu2_protocol_logic.h:58:14: fatal error: pgmspace.h: No such file or directory
```

eoyilmaz committed:
[MMU3] Fix for print getting cancelled after `MAX_RETRIES` is reached.

Fixed `MMU2::ToolChangeCommonOnce()` to prevent the print getting cancelled after the full screen error message is displayed `MAX_RETRIES` amount. With this modification, the full screen error message is shown on the `MAX_RETRIES` retry as usual, but then keeps showing the full screen message until the problem is solved, where as in the previous implementation the print is cancelled without any notification.

eoyilmaz committed:
[MMU3] Fixed `FullScreenMsg()` implementation (still needs work).

eoyilmaz commented:
> There are still build issues:
> 
> ```
> In file included from Marlin/src/feature/mmu3/mmu2.h:38,
> Compiling .pio/build/LPC1769/src/src/feature/bedlevel/ubl/ubl.cpp.o
>                  from Marlin/src/MarlinCore.cpp:235:
> Marlin/src/feature/mmu3/mmu2_protocol_logic.h:58:14: fatal error: pgmspace.h: No such file or directory
> ```

can you try fixing it, I can't replicate your setup, and all the build tests are passing.

thisiskeithb commented:
> all the build tests are passing.

None of the tests include MMU2/3 on LPC1768/9.

@rerickson1: Can you attach a .zip with your configs?

rerickson1 commented:
> > all the build tests are passing.
> 
> None of the tests include MMU2/3 on LPC1768/9.
> 
> @rerickson1: Can you attach a .zip with your configs?

https://github.com/rerickson1/Marlin/blob/rde/2.1.x_mmu3-1/Marlin/Configuration.h
https://github.com/rerickson1/Marlin/blob/rde/2.1.x_mmu3-1/Marlin/Configuration_adv.h
https://github.com/rerickson1/Marlin/blob/rde/2.1.x_mmu3-1/Marlin/src/pins/lpc1769/pins_MKS_SGEN_L_V2.h

thisiskeithb committed:
Add SKR 1.4 (LPC1768) & 1.4 Turbo (LPC1769) with MMU2 & MMU3 tests

thisiskeithb commented:
I added a couple LPC1678/9 tests with MMU2/3 enabled so the PR now fails the same way @rerickson1 mentioned on LPC1769:

```prolog
[Test LPC1769] BigTreeTech SKR 1.4 Turbo | MMU3...
In file included from Marlin/src/feature/mmu3/mmu2.h:38,
                 from Marlin/src/MarlinCore.cpp:235:
Marlin/src/feature/mmu3/mmu2_protocol_logic.h:58:14: fatal error: pgmspace.h: No such file or directory
```

Link to failed LPC1769 test: https://github.com/MarlinFirmware/Marlin/actions/runs/8038013547/job/21953498739#step:7:85

thisiskeithb commented:
It might be a good idea to rename `MMU2_SERIAL_PORT` to `MMU_SERIAL_PORT` since that setting now covers both MMU2 & MMU3.

eoyilmaz commented:
cool cool, I'll update the PR tonight 👍

eoyilmaz committed:
[MMU3] Fix for LPC1769 tests.

eoyilmaz committed:
[MMU3] Fix for LPC1769 tests (2).

eoyilmaz committed:
[MMU3] Fix for LPC1769 tests (3).

eoyilmaz committed:
[MMU3] Fix for LPC1769 tests (4).

eoyilmaz committed:
[MMU3] Fix for LPC1769 tests (5).

eoyilmaz committed:
[MMU3] Fix for LPC1769 tests (6).

eoyilmaz committed:
[MMU3] Fix for LPC1769 tests (7).

eoyilmaz committed:
[MMU3] Fix for LPC1769 tests (8).

eoyilmaz commented:
> I added a couple LPC1768/9 tests with MMU2/3 enabled so the PR now fails on LPC1769 the same way @rerickson1 mentioned:
> 
> ```prolog
> [Test LPC1769] BigTreeTech SKR 1.4 Turbo | MMU3...
> In file included from Marlin/src/feature/mmu3/mmu2.h:38,
>                  from Marlin/src/MarlinCore.cpp:235:
> Marlin/src/feature/mmu3/mmu2_protocol_logic.h:58:14: fatal error: pgmspace.h: No such file or directory
> ```
> 
> Link to failed LPC1769 test: https://github.com/MarlinFirmware/Marlin/actions/runs/8038013547/job/21953498739#step:7:85

Finally all tests are passing now.

eoyilmaz commented:
Hey @rerickson1 can you try my latest branch again, see if it is working okay for you.

thisiskeithb commented:
> do you need anything from my side to be able to merge this? I keep rebasing this relantlessly for the last couple of months now.

LPC tests and fixes weren’t added until 2-3 weeks ago and hopefully @rerickson1 can provide an update on if the changes worked.

rerickson1 commented:
1. There's a clash when the `BINARY_FILE_TRANSFER` is enabled.

```
In file included from Marlin/src/feature/../inc/MarlinConfigPre.h:56,
                 from Marlin/src/feature/binary_stream.cpp:23:
Marlin/src/feature/../inc/../../Configuration_adv.h:4526:25: error: expected unqualified-id before numeric constant
 4526 |     #define MAX_RETRIES 3
      |                         ^
Marlin/src/feature/binary_stream.h:449:67: note: in expansion of macro 'MAX_RETRIES'
  449 |   static const uint16_t PACKET_MAX_WAIT = 500, RX_TIMESLICE = 20, MAX_RETRIES = 0, VERSION_MAJOR = 0, VERSION_MINOR = 1, VERSION_PATCH = 0;
      |                                                                   ^~~~~~~~~~~
In file included from Marlin/src/feature/../inc/MarlinConfigPre.h:37,
                 from Marlin/src/feature/binary_stream.cpp:23:
Marlin/src/feature/binary_stream.h: In member function 'void BinaryStream::receive(char (&)[buffer_size])':
Marlin/src/feature/binary_stream.h:325:71: error: 'VERSION_MAJOR' was not declared in this scope
  325 |                   SERIAL_ECHOLNPGM("ss", sync, ",", buffer_size, ",", VERSION_MAJOR, ".", VERSION_MINOR, ".", VERSION_PATCH);
      |                                                                       ^~~~~~~~~~~~~
```

2. If I disable `BINARY_FILE_TRANSFER`, there is still a build issue:
```
Marlin/src/lcd/menu/menu_mmu2.cpp: In function 'void action_mmu2_reset()':
Marlin/src/lcd/menu/menu_mmu2.cpp:291:47: error: 'Hardware' is not a member of 'MMU2::MMU2::ResetForm'
  291 |       MMU2::mmu2.Reset(MMU2::MMU2::ResetForm::Hardware);
```

3. There are also a lot of warnings that need fixing:
```
In file included from Marlin/src/feature/mmu3/mmu_hw/errors_list.h:42,
                 from Marlin/src/feature/mmu3/messages.cpp:29:
Marlin/src/feature/mmu3/mmu_hw/../ultralcd.h:69:22: warning: 'const char* lcd_display_message_fullscreen_nonBlocking_P(const char*)' declared 'static' but never defined [-Wunused-function]
   69 |   static const char* lcd_display_message_fullscreen_nonBlocking_P(const char *msg);
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_log.cpp: In function 'void MMU2::LogErrorEvent_P(const char*)':
Marlin/src/feature/mmu3/mmu2_log.cpp:33:5: warning: statement is a reference, not call, to function 'SERIAL_ECHO_START' [-Waddress]
   33 |     SERIAL_ECHO_START; //! @todo Decide MMU errors on serial line
      |     ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_log.cpp:33:5: warning: statement has no effect [-Wunused-value]
Marlin/src/feature/mmu3/mmu2_log.cpp: In function 'void MMU2::LogEchoEvent_P(const char*)':
Marlin/src/feature/mmu3/mmu2_log.cpp:39:5: warning: statement is a reference, not call, to function 'SERIAL_ECHO_START' [-Waddress]
   39 |     SERIAL_ECHO_START;
      |     ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_log.cpp:39:5: warning: statement has no effect [-Wunused-value]
In file included from Marlin/src/feature/mmu3/mmu_hw/errors_list.h:42,
                 from Marlin/src/feature/mmu3/mmu2_error_converter.cpp:32:
Marlin/src/feature/mmu3/mmu_hw/../ultralcd.h:69:22: warning: 'const char* lcd_display_message_fullscreen_nonBlocking_P(const char*)' declared 'static' but never defined [-Wunused-function]
   69 |   static const char* lcd_display_message_fullscreen_nonBlocking_P(const char *msg);
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from Marlin/src/feature/mmu3/mmu2.cpp:44:
Marlin/src/feature/mmu3/mmu2.cpp: In member function 'void MMU2::MMU2::SetCurrentTool(uint8_t)':
Marlin/src/feature/mmu3/mmu2_log.h:72:15: warning: statement is a reference, not call, to function 'SERIAL_ECHO_START' [-Waddress]
   72 |               SERIAL_ECHO_START;   \
      |               ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2.cpp:560:5: note: in expansion of macro 'MMU2_ECHO_MSGRPGM'
  560 |     MMU2_ECHO_MSGRPGM(PSTR("MMU2tool="));
      |     ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_log.h:72:15: warning: statement has no effect [-Wunused-value]
   72 |               SERIAL_ECHO_START;   \
      |               ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2.cpp:560:5: note: in expansion of macro 'MMU2_ECHO_MSGRPGM'
  560 |     MMU2_ECHO_MSGRPGM(PSTR("MMU2tool="));
      |     ^~~~~~~~~~~~~~~~~
In file included from Marlin/src/feature/mmu3/mmu_hw/errors_list.h:42,
                 from Marlin/src/feature/mmu3/mmu2_progress_converter.cpp:34:
Marlin/src/feature/mmu3/mmu_hw/../ultralcd.h:69:22: warning: 'const char* lcd_display_message_fullscreen_nonBlocking_P(const char*)' declared 'static' but never defined [-Wunused-function]
   69 |   static const char* lcd_display_message_fullscreen_nonBlocking_P(const char *msg);
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp:30:
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp: In member function 'void MMU2::ProtocolLogic::LogRequestMsg(const uint8_t*, uint8_t)':
Marlin/src/feature/mmu3/mmu2_log.h:65:15: warning: statement is a reference, not call, to function 'SERIAL_ECHO_START' [-Waddress]
   65 |               SERIAL_ECHO_START;   \
      |               ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp:799:7: note: in expansion of macro 'MMU2_ECHO_MSGLN'
  799 |       MMU2_ECHO_MSGLN(tmp);
      |       ^~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_log.h:65:15: warning: statement has no effect [-Wunused-value]
   65 |               SERIAL_ECHO_START;   \
      |               ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp:799:7: note: in expansion of macro 'MMU2_ECHO_MSGLN'
  799 |       MMU2_ECHO_MSGLN(tmp);
      |       ^~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp: In member function 'void MMU2::ProtocolLogic::LogError(const char*)':
Marlin/src/feature/mmu3/mmu2_log.h:72:15: warning: statement is a reference, not call, to function 'SERIAL_ECHO_START' [-Waddress]
   72 |               SERIAL_ECHO_START;   \
      |               ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_log.h:76:35: note: in expansion of macro 'MMU2_ECHO_MSGRPGM'
   76 |     #define MMU2_ERROR_MSGRPGM(S) MMU2_ECHO_MSGRPGM(S) //! @todo Decide MMU errors  on serial line
      |                                   ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp:808:5: note: in expansion of macro 'MMU2_ERROR_MSGRPGM'
  808 |     MMU2_ERROR_MSGRPGM(reason_P);
      |     ^~~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_log.h:72:15: warning: statement has no effect [-Wunused-value]
   72 |               SERIAL_ECHO_START;   \
      |               ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_log.h:76:35: note: in expansion of macro 'MMU2_ECHO_MSGRPGM'
   76 |     #define MMU2_ERROR_MSGRPGM(S) MMU2_ECHO_MSGRPGM(S) //! @todo Decide MMU errors  on serial line
      |                                   ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp:808:5: note: in expansion of macro 'MMU2_ERROR_MSGRPGM'
  808 |     MMU2_ERROR_MSGRPGM(reason_P);
      |     ^~~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp: In member function 'void MMU2::ProtocolLogic::LogResponse()':
Marlin/src/feature/mmu3/mmu2_log.h:65:15: warning: statement is a reference, not call, to function 'SERIAL_ECHO_START' [-Waddress]
   65 |               SERIAL_ECHO_START;   \
      |               ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp:816:5: note: in expansion of macro 'MMU2_ECHO_MSGLN'
  816 |     MMU2_ECHO_MSGLN(lrb);
      |     ^~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_log.h:65:15: warning: statement has no effect [-Wunused-value]
   65 |               SERIAL_ECHO_START;   \
      |               ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp:816:5: note: in expansion of macro 'MMU2_ECHO_MSGLN'
  816 |     MMU2_ECHO_MSGLN(lrb);
      |     ^~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp: In member function 'void MMU2::ProtocolLogic::SendMsg(modules::protocol::RequestMsg)':
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp:242:21: warning: 'i' is used uninitialized in this function [-Wuninitialized]
  242 |       for ( uint8_t i; i < len; i++) {
      |                     ^
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp: In member function 'void MMU2::ProtocolLogic::SendWriteMsg(modules::protocol::RequestMsg)':
Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp:262:21: warning: 'i' is used uninitialized in this function [-Wuninitialized]
  262 |       for ( uint8_t i; i < len; i++) {
      |                     ^
In file included from Marlin/src/feature/mmu3/mmu2_reporting.cpp:41:
Marlin/src/feature/mmu3/mmu2_reporting.cpp: In member function 'void MMU2::TryLoadUnloadReporter::DumpToSerial()':
Marlin/src/feature/mmu3/mmu2_log.h:65:15: warning: statement is a reference, not call, to function 'SERIAL_ECHO_START' [-Waddress]
   65 |               SERIAL_ECHO_START;   \
      |               ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_reporting.cpp:562:5: note: in expansion of macro 'MMU2_ECHO_MSGLN'
  562 |     MMU2_ECHO_MSGLN(buf);
      |     ^~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_log.h:65:15: warning: statement has no effect [-Wunused-value]
   65 |               SERIAL_ECHO_START;   \
      |               ^~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/mmu2_reporting.cpp:562:5: note: in expansion of macro 'MMU2_ECHO_MSGLN'
  562 |     MMU2_ECHO_MSGLN(buf);
      |     ^~~~~~~~~~~~~~~
In file included from Marlin/src/feature/mmu3/mmu_hw/errors_list.h:42,
                 from Marlin/src/feature/mmu3/mmu2_reporting.cpp:49:
Marlin/src/feature/mmu3/mmu_hw/../ultralcd.h: At global scope:
Marlin/src/feature/mmu3/mmu_hw/../ultralcd.h:69:22: warning: 'const char* lcd_display_message_fullscreen_nonBlocking_P(const char*)' declared 'static' but never defined [-Wunused-function]
   69 |   static const char* lcd_display_message_fullscreen_nonBlocking_P(const char *msg);
      |                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Marlin/src/feature/mmu3/ultralcd.cpp: In function 'const char* lcd_display_message_fullscreen_nonBlocking_P(const char*)':
Marlin/src/feature/mmu3/ultralcd.cpp:114:10: warning: variable 'multi_screen' set but not used [-Wunused-but-set-variable]
  114 |     bool multi_screen = false;
      |          ^~~~~~~~~~~~
Marlin/src/inc/Warnings.cpp:702:6: warning: #warning "Motherboard DIAG jumpers must be removed when SENSORLESS_HOMING is disabled. (Define DIAG_JUMPERS_REMOVED to suppress this warning.)" [-Wcpp]
  702 |     #warning "Motherboard DIAG jumpers must be removed when SENSORLESS_HOMING is disabled. (Define DIAG_JUMPERS_REMOVED to suppress this warning.)"
      |      ^~~~~~~
In file included from Marlin/src/lcd/menu/../../inc/../HAL/../HAL/LPC1768/../shared/Marduino.h:36,
                 from Marlin/src/lcd/menu/../../inc/../HAL/../HAL/LPC1768/HAL.h:37,
                 from Marlin/src/lcd/menu/../../inc/../HAL/HAL.h:30,
                 from Marlin/src/lcd/menu/../../inc/MarlinConfig.h:33,
                 from Marlin/src/lcd/menu/menu_mmu2.cpp:23:
Marlin/src/lcd/menu/menu_mmu2.cpp: In function 'void menu_mmu2_toolchange_stat_total()':
Marlin/src/lcd/menu/menu_mmu2.cpp:240:27: warning: format '%u' expects argument of type 'unsigned int', but argument 3 has type 'uint32_t' {aka 'long unsigned int'} [-Wformat=]
  240 |   sprintf_P(buffer2, PSTR("%u"), MMU2::operation_statistics.tool_change_total_counter);
/Users/ryan/.platformio/packages/framework-arduino-lpc176x/cores/arduino/Arduino.h:58:18: note: in definition of macro 'PSTR'
   58 | #define PSTR(v) (v)
```

rerickson1 commented:
My changes are rebased onto this PR: https://github.com/rerickson1/Marlin/compare/2efbcbf529b02d2a25079712f63dae1a4af2f97c...0d12d84160a6df96caf22cbace0150ecd28d98c5

My branch: https://github.com/rerickson1/Marlin/tree/rde/2.1.x_mmu3-2

eoyilmaz commented:
@rerickson1 for the "binary file transfer" feature, from the error message I see that the binary file transfer feature is also defining the `MAX_RETRIES`, we can rename the variable to `MMU_MAX_RETRIES`

and I don't think warnings are important...

eoyilmaz committed:
[MMU3] Renamed `MAX_RETRIES` to `MMU2_MAX_RETRIES`.

eoyilmaz committed:
[MMU3] Fixed `lcd.menu.menu_mmu2.action_mmu2_reset()`.

It was errorenously calling `MMU2::MMU2::ResetForm::Hardware` which doesn't exist replaced with `ResetForm::ResetPin`.

eoyilmaz commented:
@rerickson1 good catch! The following should also be fixed now:

```shell
Marlin/src/lcd/menu/menu_mmu2.cpp: In function 'void action_mmu2_reset()':
Marlin/src/lcd/menu/menu_mmu2.cpp:291:47: error: 'Hardware' is not a member of 'MMU2::MMU2::ResetForm'
  291 |       MMU2::mmu2.Reset(MMU2::MMU2::ResetForm::Hardware);
```

rerickson1 commented:
> and I don't think warnings are important...

I have to disagree. Compiler warnings are there for a reason. When I build the firmware for my board the only warnings that exist are deliberate ones to inform the user of something important.
A good codebase should not have C/C++ based warnings anywhere when compiling.
The tests run should catch warnings and fail the PR until they are fixed.

This is my opinion so I will leave it up to the project maintainers to enforce.

rerickson1 commented:
Good news, it compiles now!
This will allow me to get the new MMU2 firmware installed and then I can try this out!

Here are the warnings I get when compiling:
[mmu3_warnings.log](https://github.com/MarlinFirmware/Marlin/files/14641064/mmu3_warnings.log)

rerickson1 commented on Marlin/src/feature/mmu3/mmu2_log.h:
```diff
@@ -0,0 +1,96 @@
+
+    #define MMU2_ECHO_MSGLN(S)   \
+            do {                     \
+              SERIAL_ECHO_START;   \
```
A perfect example of build warnings pointing out a problem:

```suggestion
              SERIAL_ECHO_START();   \
```

rerickson1 commented on Marlin/src/feature/mmu3/mmu2_log.cpp:
```diff
@@ -0,0 +1,45 @@
+  namespace MMU2 {
+
+  void LogErrorEvent_P(const char *msg) {
+    SERIAL_ECHO_START; //! @todo Decide MMU errors on serial line
```
```suggestion
    SERIAL_ECHO_START(); //! @todo Decide MMU errors on serial line
```

rerickson1 commented on Marlin/src/feature/mmu3/ultralcd.h:
```diff
@@ -0,0 +1,76 @@
+
+  bool pgm_is_whitespace(const char *c_addr);
+  bool pgm_is_interpunction(const char *c_addr);
+  static const char* lcd_display_message_fullscreen_nonBlocking_P(const char *msg);
```
This function is not declared anywhere.

```suggestion
```

rerickson1 commented on Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp:
```diff
@@ -0,0 +1,924 @@
+
+    #if defined(__AVR__) || defined(TARGET_LPC1768)
+      // TODO: I'm not sure if this is the correct approach with AVR
+      for ( uint8_t i; i < len; i++) {
```
```suggestion
      for ( uint8_t i = 0; i < len; i++) {
```

rerickson1 commented on Marlin/src/feature/mmu3/mmu2_protocol_logic.cpp:
```diff
@@ -0,0 +1,924 @@
+    uint8_t len = Protocol::EncodeRequest(rq, txbuff);
+    #if defined(__AVR__) || defined(TARGET_LPC1768)
+      // TODO: I'm not sure if this is the correct approach with AVR
+      for ( uint8_t i; i < len; i++) {
```
```suggestion
      for ( uint8_t i = 0; i < len; i++) {
```

rerickson1 commented on Marlin/src/feature/mmu3/ultralcd.cpp:
```diff
@@ -0,0 +1,225 @@
+   */
+  static const char* lcd_display_message_fullscreen_nonBlocking_P(const char *msg) {
+    const char *msgend = msg;
+    bool multi_screen = false;
```
This variable is assigned but never used. Remove it and get rid of commented out code thats using it.

```suggestion
```

rerickson1 commented on Marlin/src/lcd/menu/menu_mmu2.cpp:
```diff
@@ -92,23 +134,218 @@ void menu_mmu2_eject_filament() {
+  sprintf_P(buffer1, PSTR("%u"), MMU2::operation_statistics.tool_change_counter);
+
+  char buffer2[LCD_WIDTH];
+  sprintf_P(buffer2, PSTR("%u"), MMU2::operation_statistics.tool_change_total_counter);
```
```suggestion
  sprintf_P(buffer2, PSTR("%lu"), MMU2::operation_statistics.tool_change_total_counter);
```

rerickson1 reviewed (changes requested):
This patch fixes all warnings.

[Changes-On-b3d58d.patch](https://github.com/MarlinFirmware/Marlin/files/14641217/Changes-On-b3d58d.patch)

eoyilmaz commented:
@rerickson1 thank you for your contribution, I should have said "the warnings are not important **at this stage**". And the only warnings that I'm getting are these:

```shell
Marlin/src/inc/Warnings.cpp:223:6: warning: #warning "Note: Auto-assigned Z2 STEP/DIR/ENABLE_PINs to unused En_STEP/DIR/ENABLE_PINs. (Define NO_AUTO_ASSIGN_WARNING to suppress this warning.)" [-Wcpp]
  223 |     #warning "Note: Auto-assigned Z2 STEP/DIR/ENABLE_PINs to unused En_STEP/DIR/ENABLE_PINs. (Define NO_AUTO_ASSIGN_WARNING to suppress this warning.)"
      |      ^~~~~~~
Marlin/src/inc/Warnings.cpp:702:6: warning: #warning "Motherboard DIAG jumpers must be removed when SENSORLESS_HOMING is disabled. (Define DIAG_JUMPERS_REMOVED to suppress this warning.)" [-Wcpp]
  702 |     #warning "Motherboard DIAG jumpers must be removed when SENSORLESS_HOMING is disabled. (Define DIAG_JUMPERS_REMOVED to suppress this warning.)"
      |      ^~~~~~~
```

I believe your configuration is triggering or disabling some code paths that I don't normally run. I'll include your patch thank you 👍

eoyilmaz committed:
[MMU3] Fixed warnings (contributed by @rerickson1).

eoyilmaz commented:
@rerickson1 okay I included your changes, please check again if you are getting any more errors.

rerickson1 commented:
@eoyilmaz all warnings gone and now build errors!

thinkyhead commented:
> do you need anything from my side to be able to merge this? I keep rebasing this relantlessly for the last couple of months now.

In recent days I've been catching up on PRs after a period of reduced activity (i.e., mental health break) in prep for another big sprint towards the next release. I'll be sure to get this reviewed, adjusted, and likely merged very soon.

thinkyhead commented:
How important is it to maintain parity between this code and Prusa firmwares? I would like to remove the MMU2 namespace and apply Marlin coding standards, but if there is a need to maintain this in parallel with Prusa firmware that complicates things a bit.

There are thousands of very particular changes that I would like to make to this code to bring it into alignment with our coding style and standards. Of course I cannot ask you to make all those changes yourself, as it would take forever. So I will simply apply changes in commits as time allows until we finally get to a point where everything is proper. Among the things I would like to fix:
- No `#if HAS_PRUSA_MMU3` wrappers in `.h` files. (The `.h` file is simply not included when `!HAS_PRUSA_MMU3`.)
- Get rid of the "MMU2" namespace (maybe replacing it with a class wrapper).
- Rename stuff that pertains to the MMU3 as "MMU3…" instead of "MMU2…" naming.
- Replace `MMU2_MENUS` with `MMU_MENUS` … since it doesn't just pertain to MMU2.
- Use common naming instead of maintaining multiple names, such as `load_filament` / `load_filament_to_nozzle` and `load_filament` / `load_to_feeder`.
- …and much more.

There is some likelihood that things will get broken as I attempt to merely clean up the code without affecting behavior. So as changes are made you will need to verify that nothing has been broken in the process.

This process may take several weeks, so I beg your further patience as we work through all the details.

thinkyhead committed:
MMU2_MENUS => MMU_MENUS

thinkyhead committed:
initial review cleanup

thinkyhead committed:
fix CPP wrapping

thinkyhead committed:
fix mmu log

thinkyhead committed:
and so on…

thinkyhead committed:
Towards proper string types

eoyilmaz commented:
Hey @thinkyhead thanks for looking in to the PR 👍 

> How important is it to maintain parity between this code and Prusa firmwares?

I'm monitoring Prusa-Firmware-MMU project and I think we can diverge as much as we like. No important to keep parity. The changes are mostly one liners, which can easily be transfered.

> Of course I cannot ask you to make all those changes yourself, as it would take forever.

With supervision, I'm okay to work on the code as much as I can. But I'm not mainly a C/C++ developer (not at least for the last 10-12 years). Yeah, with your supervision I would like to keep working on the code.

> This process may take several weeks, so I beg your further patience as we work through all the details.

No worries, I'm all in, sorry if I sound as if I'm complaining, the PR was sitting there for a long time, but again I'm ready to spend as much time and effort as required if we keep progressing 👍

thinkyhead committed:
move translatable strings

thinkyhead committed:
strings followup

thinkyhead committed:
comment updates

eoyilmaz commented:
Hey @thinkyhead I tested the code with your changes, it seems to be all working fine right now 👍

thisiskeithb commented on Marlin/Configuration_adv.h:
```diff
@@ -4378,44 +4378,89 @@
-  // Add an LCD menu for MMU2
-  //#define MMU2_MENUS
+  // Add an LCD menu for MMU2/MMU2S/MMU3
+  //#define MMU_MENUS
```
This rename will need to be added to `Changes.h`.

thinkyhead committed:
optimize SpoolJoin singleton

thinkyhead committed:
Optimize CRC8 singleton

thinkyhead committed:
misc. formatting

thinkyhead committed:
lowercase Protocol class methods

thinkyhead committed:
save for later

thinkyhead committed:
Apply some standards

thinkyhead committed:
Update Changes.h

thinkyhead committed:
match mmu2 naming

thinkyhead committed:
MMU2 => 3

thinkyhead committed:
fix overlap

thinkyhead committed:
fix build

microtronics commented:
I really like the progress that has been made here and can't wait to see it upstream. If I have the time I'll also give the code on @eoyilmaz branch a try.
However, are you aware that Prusa is finally shipping the MMU3 for MK4 also? The firmware includes a new ramming sequence and is based on Marlin2, afaik:
https://github.com/prusa3d/Prusa-Firmware-Buddy/releases/tag/v6.0.0-RC1
On the other hand, some of this code seems to make use of the load cell which is likely not a commonly used feature in the wild...

Currently, the latest MMU3 firmware is not backwards compatible to the MK3 anymore:
https://github.com/prusa3d/Prusa-Firmware-MMU/releases/tag/v3.0.3
I'm not sure if this will be fixed.

rerickson1 commented:
@eoyilmaz I got this running on my printer now and updated my MMU2S to v3.0.2 firmware. So far so good. I haven't done any multi-color prints yet. I can load and unload a single filament reliably though.
One issue I do see though. As soon as the print starts my LCD is freezing. Would love to get that issue solved. Its been an issue ever since I started using the MMU2. Once the print finishes, the LCD starts working again.
Video provided in the issue here: https://github.com/MarlinFirmware/Marlin/issues/26544

eoyilmaz commented:
@rerickson1 the freezing issue is old #16495 I mean I had it even MMU3 was not a thing, and I couldn't find the source.

rerickson1 commented:
> @rerickson1 the freezing issue is old #16495 I mean I had it even MMU3 was not a thing, and I couldn't find the source.

Thanks for clarification. I thought you may have fixed it during this PR.
So your LCD still freezes too when using this PR?

rerickson1 commented:
@eoyilmaz I am hoping you can help me tune some of my firmware values.

I am using a Bondtech BMG and E3D Revo Six.

![image](https://github.com/MarlinFirmware/Marlin/assets/4430317/65c69f92-c606-4e18-861a-d52a5853dbb0)

Based on the measurements above, what should my values be for:
`MMU2_EXTRUDER_PTFE_LENGTH` = ?
`MMU2_EXTRUDER_HEATBREAK_LENGTH` = ?
`MMU2_LOAD_TO_NOZZLE_LENGTH` = 87 + 5 <- This looks good

Then can you explain when `MMU2_TOOL_CHANGE_LOAD_LENGTH` and `MMU2_CHECK_FILAMENT_PRESENCE_EXTRUSION_LENGTH` are used?

eoyilmaz commented:
Hey @rerickson1, I'm using a Creality Sprite Extruder Pro and my settings are as follows:

```cpp
MMU2_EXTRUDER_PTFE_LENGTH        5.0
MMU2_EXTRUDER_HEATBREAK_LENGTH  20.0
MMU2_LOAD_TO_NOZZLE_LENGTH 87.0 + 5.0
```

And my slicer settings are:

![Screenshot 2024-04-12 at 13 16 08](https://github.com/MarlinFirmware/Marlin/assets/1786804/2e9b71f1-4c16-4302-8128-6025ed119726)

So, in real life the PTFE tube length of Sprite Extruder Pro is 19 mm, but I set it to 5 mm in the slicer, 5 mm was the default for MK3.5 in the Prusa Slicer which has a physical length of 24 mm if I'm not mistaken, and if I set it to any length longer than this I have huge issues with filament tips getting thicker and thicker and cause clogging after 50 or so tool changes. But with 5 mm, I get brialliant tips and I can easily go thousands of tool changes without intervention. And I keep the Cooling Tube Position at 45 mm which is the center of the PTFE tube, and this matches the physically center of the PTFE tube inside the extruder. Oh also I replace the PTFE tube around 10k tool changes or whenever I start having ticker filament tips.

with your extruder I would set the Cooling Tube Position to 51 mm and the Cooling tube length to 5 mm in the slicer.

Btw, I didn't try using the new Stamping feature with Prusa Slicer 2.7.3 yet...

rerickson1 commented:
Thanks @eoyilmaz for the info! I am working on tuning settings and found something interesting.
I am using the "Load to nozzle" command from the LCD menu. I have noticed it behaves differently if I have not performed an auto home first. Any ideas on why this might be? Maybe there should be something that homes all axis before MMU commands can happen?

rerickson1 commented:
I believe I have found another issue. Reproduce by:
1. Start printer
2. Auto home all
3. Issue T4 command from the terminal to load filament (I have filament in the 5th slot for testing)
From this point the MMU unit does not try to load the filament to the nozzle. It just keeps trying to purge in a loop.
Previously I could issue Tx command to load any filament.

rerickson1 commented:
Also, error messages not displayed correctly.
![IMG_4727](https://github.com/MarlinFirmware/Marlin/assets/4430317/6f74a5c2-6e5f-41d7-b0b8-8d32e53a3b71)

thisiskeithb commented:
Note: Lots of collaboration was buried, so be sure to click the `Load more...` link above:

<img width="213" alt="image" src="https://github.com/MarlinFirmware/Marlin/assets/13375512/a64de21e-b897-419e-9988-87fcfa5c86d5">

Minims commented:
Hello, is this PR is still a WIP or it will be integrated soon as new Feature on Marlin ?
Maybe as a BETA / EXPERIMENTAL feature and some smaller PR will help to fix issues from new comers ?

eoyilmaz commented:
I keep rebasing this... until it is merged.

SpankyDevil commented:
@eoyilmaz  Are we any closer to getting this merged?

thisiskeithb commented:
> @eoyilmaz Are we any closer to getting this merged?

Merging will be up to thinkyhead.

thisiskeithb commented:
I've rebased this PR to bring it up to date with current `bugfix-2.1.x` & resolved merge conflicts, so please update your local copy with the following before pushing additional commits:

```shell
git checkout bugfix-2.1.x_mmu3_pr
git fetch origin
git reset --hard origin/bugfix-2.1.x_mmu3_pr
```

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into pr/26635

thinkyhead committed:
cleanup

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.1.x' into pr/26635

thinkyhead committed:
cleanup after merge

thinkyhead commented:
I'll have this merged shortly for wider testing. Feel free to submit updates to clean up or improve anything, or to add MMU3 menu items for other LCD controllers. In general any values stored in EEPROM that are associated with a G-code should be reported as such so that users can replay that G-code to recover settings. And, if MMU3 settings exist which are stored to EEPROM but which don't have an associated G-code, then a G-code and/or parameters should be created for each of those settings. We will then document those G-codes and other firmwares can choose to adopt them, or not.

thinkyhead committed:
namespace
